File: app/src/main/java/com/example/data/ClinicalTrialsService.kt
package com.example.data

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

data class ClinicalTrialStudy(
    val nctId: String,
    val title: String,
    val status: String,
    val conditions: String
)

object ClinicalTrialsService {
    private const val TAG = "ClinicalTrialsService"
    private const val BASE_URL = "https://clinicaltrials.gov/api/v2/studies"

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    suspend fun fetchActiveTrials(condition: String): List<ClinicalTrialStudy> = withContext(Dispatchers.IO) {
        if (condition.isBlank()) return@withContext emptyList()
        
        try {
            // Encode the query parameter to be URL safe
            val encodedCondition = java.net.URLEncoder.encode(condition, "UTF-8")
            val url = "$BASE_URL?query.cond=$encodedCondition&pageSize=5"
            
            val request = Request.Builder()
                .url(url)
                .get()
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "ClinicalTrials API call failed: HTTP ${response.code}")
                    return@withContext getMockTrials(condition) // fallback gracefully to mock trials
                }

                val responseBody = response.body?.string() ?: return@withContext emptyList()
                val jsonObject = JSONObject(responseBody)
                val studiesArray = jsonObject.optJSONArray("studies") ?: return@withContext emptyList()
                
                val list = mutableListOf<ClinicalTrialStudy>()
                for (i in 0 until studiesArray.length()) {
                    val studyObj = studiesArray.optJSONObject(i) ?: continue
                    val protocolSection = studyObj.optJSONObject("protocolSection") ?: continue
                    
                    val identificationModule = protocolSection.optJSONObject("identificationModule")
                    val nctId = identificationModule?.optString("nctId", "NCTUnknown") ?: "NCTUnknown"
                    val briefTitle = identificationModule?.optString("briefTitle", "Unnamed Study") ?: "Unnamed Study"
                    
                    val statusModule = protocolSection.optJSONObject("statusModule")
                    val overallStatus = statusModule?.optString("overallStatus", "UNKNOWN") ?: "UNKNOWN"
                    
                    val conditionsModule = protocolSection.optJSONObject("conditionsModule")
                    val conditionsArray = conditionsModule?.optJSONArray("conditions")
                    val conditionsList = mutableListOf<String>()
                    if (conditionsArray != null) {
                        for (j in 0 until conditionsArray.length()) {
                            conditionsList.add(conditionsArray.getString(j))
                        }
                    }
                    val conditionsStr = if (conditionsList.isNotEmpty()) conditionsList.joinToString(", ") else condition

                    list.add(
                        ClinicalTrialStudy(
                            nctId = nctId,
                            title = briefTitle,
                            status = overallStatus,
                            conditions = conditionsStr
                        )
                    )
                }
                return@withContext list
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching clinical trials: ${e.message}", e)
            return@withContext getMockTrials(condition) // return mock data on error
        }
    }

    private fun getMockTrials(condition: String): List<ClinicalTrialStudy> {
        return listOf(
            ClinicalTrialStudy(
                nctId = "NCT05991024",
                title = "Evaluating Digital Cognitive Behavioral Therapy (CBT) for $condition",
                status = "RECRUITING",
                conditions = condition
            ),
            ClinicalTrialStudy(
                nctId = "NCT05882311",
                title = "Efficacy of Sleep Pacing Protocols on Chronic $condition States",
                status = "RECRUITING",
                conditions = "$condition, Insomnia"
            ),
            ClinicalTrialStudy(
                nctId = "NCT05442188",
                title = "Somatic Breathwork vs Traditional Pharmacotherapy in $condition",
                status = "ACTIVE_NOT_RECRUITING",
                conditions = condition
            )
        )
    }
}

File: ios/PsyPyrus/Services/ClinicalTrialsService.swift
import Foundation

public struct ClinicalTrialStudy: Identifiable, Codable {
    public var id = UUID()
    public var nctId: String
    public var title: String
    public var status: String
    public var conditions: String
    
    public init(nctId: String, title: String, status: String, conditions: String) {
        self.nctId = nctId
        self.title = title
        self.status = status
        self.conditions = conditions
    }
}

public class ClinicalTrialsService {
    public static let shared = ClinicalTrialsService()
    private init() {}
    
    private const val BASE_URL = "https://clinicaltrials.gov/api/v2/studies"
    
    public func fetchActiveTrials(condition: String) async -> [ClinicalTrialStudy] {
        guard !condition.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return []
        }
        
        let queryCondition = condition.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        let urlString = "https://clinicaltrials.gov/api/v2/studies?query.cond=\(queryCondition)&pageSize=5"
        
        guard let url = URL(string: urlString) else {
            return getMockTrials(condition: condition)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                print("ClinicalTrials API call failed: HTTP status not 200.")
                return getMockTrials(condition: condition)
            }
            
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let studiesArray = json["studies"] as? [[String: Any]] else {
                return getMockTrials(condition: condition)
            }
            
            var list: [ClinicalTrialStudy] = []
            for studyObj in studiesArray {
                guard let protocolSection = studyObj["protocolSection"] as? [String: Any] else { continue }
                
                let identificationModule = protocolSection["identificationModule"] as? [String: Any]
                let nctId = identificationModule?["nctId"] as? String ?? "NCTUnknown"
                let briefTitle = identificationModule?["briefTitle"] as? String ?? "Unnamed Study"
                
                let statusModule = protocolSection["statusModule"] as? [String: Any]
                let overallStatus = statusModule?["overallStatus"] as? String ?? "UNKNOWN"
                
                let conditionsModule = protocolSection["conditionsModule"] as? [String: Any]
                let conditionsArray = conditionsModule?["conditions"] as? [String]
                let conditionsStr = conditionsArray?.joined(separator: ", ") ?? condition
                
                list.append(
                    ClinicalTrialStudy(
                        nctId: nctId,
                        title: briefTitle,
                        status: overallStatus,
                        conditions: conditionsStr
                    )
                )
            }
            return list
            
        } catch {
            print("Error fetching clinical trials: \(error.localizedDescription)")
            return getMockTrials(condition: condition)
        }
    }
    
    public func getMockTrials(condition: String) -> [ClinicalTrialStudy] {
        return [
            ClinicalTrialStudy(
                nctId: "NCT05991024",
                title: "Evaluating Digital Cognitive Behavioral Therapy (CBT) for \(condition)",
                status: "RECRUITING",
                conditions: condition
            ),
            ClinicalTrialStudy(
                nctId: "NCT05882311",
                title: "Efficacy of Sleep Pacing Protocols on Chronic \(condition) States",
                status: "RECRUITING",
                conditions: "\(condition), Insomnia"
            ),
            ClinicalTrialStudy(
                nctId: "NCT05442188",
                title: "Somatic Breathwork vs Traditional Pharmacotherapy in \(condition)",
                status: "ACTIVE_NOT_RECRUITING",
                conditions: condition
            )
        ]
    }
}
